Data converter for GEF-CPT files

antbrown opened the pull request:
## Description

Please see https://github.com/SeequentEvo/evo-data-converters/issues/78 for what this Pull Request sets out to do.

On the technical side of things:

- We use pygef to parse and read a collection of GEF-CPT files into a data structure of `{"hole-001": CPTData, "hole-002": CPTData}`
- We then turn this into a common Downhole Collection intermediary object using standard pandas dataframes for the hole collars information and each measurement table (for GEF there is only one and it is a distance table)
- There are table adapters to handle the different types of measurement dataframes (interval and distance)- this will become useful for converting AGS in the future
- A common DownholeCollection -> Evo GeoscienceObject converter takes the intermediary object and turns it into an Evo object, including attributes on the collars and attributes on the measurement tables- supporting booleans, dates, integers, floats, strings and categorical attributes currently

## Checklist

- [x] I have read the contributing guide and the code of conduct

reflecting-moon committed:
Data converter for GEF-CPT files

Co-authored-by: PaulW8787 <182706494+PaulW8787@users.noreply.github.com>
Co-authored-by: antbrown <1214401+antbrown@users.noreply.github.com>

l-macdonald commented:
@antbrown could you please update the codeowners file, assigning the `gef-maintainers` group to the gef package thanks.

PaulW8787 committed:
AttributeDescription records if pandas pint information available

- Converts from PintType to Evo unit types
- [GEF] Adds Pint measurement units to the measurement dataframe using GEF column name as unit mapping lookup.

Co-authored-by: GaryWL <gary@catalyst.net.nz>

daniel-kinney-seequent commented on packages/gef/pyproject.toml:
```diff
@@ -0,0 +1,60 @@
+name = "evo-data-converters-gef"
+description = "Python data converters for GEF to Evo geoscience objects"
+version = "0.0.1"
+requires-python = ">=3.10"
```
Should this be >=3.11, because of pygef?

antbrown replied:
The bulk of the features will work on python 3.10, so I've added a note to the README to say that python 3.11+ is required to ingest the extra headers.

daniel-kinney-seequent commented on packages/gef/README.md:
```diff
@@ -0,0 +1,94 @@
+
+## Pre-requisites
+
+* Python virtual environment with Python 3.10, 3.11, or 3.12
```
I think you need to remove 3.10?

daniel-kinney-seequent commented on packages/common/src/evo/data_converters/common/objects/attributes.py:
```diff
@@ -0,0 +1,266 @@
+import pyarrow as pa
+import typing
+from enum import Enum
+from dataclasses import dataclass
```
The linter/formatter is not complaining, but to my eye the imports are in a strange order. What I expect to see is:

# Standard library imports first
# Third party imports seconds
# evo-data-converter common imports third
# gef imports last

And within each section, I would expect them to be ordered alphabetically. It seems like the sort of thing our formatter should take care of.

antbrown replied:
Yeah, I noticed this too - the `I` rule isn't being enforced even though it is selected in `pyproject.toml`

```
[tool.ruff.lint]
select = ["E4", "E7", "E9", "F", "I", "UP006"]
```

I'll investigate further, and if I can't get it working I'll just run it manually with `uv run ruff check --select I --fix packages/gef`

daniel-kinney-seequent commented on packages/common/src/evo/data_converters/common/objects/attributes.py:
```diff
@@ -0,0 +1,266 @@
+    """Factory for creating PyArrow tables from pandas Series with specified data types."""
+
+    @staticmethod
+    def create_table(series: pd.Series, data_type: DataType) -> pa.Table:
```
I'm not sure why this would be a class/staticmethod as opposed to a free function?

Gary-Lockyer replied:
Done

daniel-kinney-seequent commented on packages/common/src/evo/data_converters/common/objects/attributes.py:
```diff
@@ -0,0 +1,266 @@
+            ),
+        )
+
+        lookup_table_args = client.save_table(lookup_table)
```
This is significant file IO and I'm surprised it's not an async function. Anway, just a note - nothing to do about.

daniel-kinney-seequent commented on packages/common/src/evo/data_converters/common/objects/attributes.py:
```diff
@@ -0,0 +1,266 @@
+    }
+
+    @staticmethod
+    def create(name: str, series: pd.Series, client: ObjectDataClient) -> OneOfAttribute_Item | None:
```
Where is this used?

daniel-kinney-seequent commented on packages/gef/src/evo/data_converters/gef/converter/parse_gef_files.py:
```diff
@@ -0,0 +1,135 @@
+            return [cpt_data]
+
+        else:
+            raise ValueError(f"File '{filepath}' has extension '{ext}', expected .xml or .gef.")
```
Let's avoid raising here. Log an error, skip it, and carry on. I wrote this in a couple places, but I'm directing that comment in general at the parsing code. I don't think the whole import should fail because of a single bad file.

daniel-kinney-seequent replied:
@antbrown convinced me it's okay as it is.

daniel-kinney-seequent commented on packages/gef/src/evo/data_converters/gef/converter/gef_to_downhole_collection.py:
```diff
@@ -0,0 +1,439 @@
+            raise ValueError(f"CPT file '{hole_id}' has invalid EPSG code in SRS name: '{srs_name}'. Error: {e}")
+
+        if epsg_code == 404000:
+            epsg_code = "unspecified"
```
It's weird to return a string here. Maybe a `None`.

daniel-kinney-seequent commented on packages/gef/src/evo/data_converters/gef/converter/gef_to_downhole_collection.py:
```diff
@@ -0,0 +1,439 @@
+                items = cpt_data.raw_headers[raw_header_name]
+
+                for id_, *values in items:
+                    value = ", ".join(str(v) for v in values)
```
I'd like to see an example of when we would get a comma separated list.

daniel-kinney-seequent commented on packages/gef/src/evo/data_converters/gef/converter/gef_to_downhole_collection.py:
```diff
@@ -0,0 +1,439 @@
+        }
+        collar_attributes = self._get_collar_attributes(cpt_data)
+        raw_header_info = self._get_raw_header_info(cpt_data)
+        return collar_data | collar_attributes | raw_header_info
```
I'm concerned about key collision, especially with `_get_collar_attributes` ripping through `var(cpt_data)`.

PaulW8787 replied:
Looking through the code, this won't happen. There are no possible key overlaps that I can see. However this could be ensured by writing a dict-merge method which raises an exception in the case of overlapping keys. Leaving that to a future iteration.

daniel-kinney-seequent commented on packages/gef/src/evo/data_converters/gef/converter/gef_to_downhole_collection.py:
```diff
@@ -0,0 +1,439 @@
+
+        :return: Pandas DataFrame with typed collar data
+        """
+        return pd.DataFrame(self.collar_rows).astype(dtype=self.COLLAR_DTYPES)
```
How does that work with the dtype? Can't there be a lot more columns than what's in `self.COLLAR_DTYPES`?

antbrown replied:
correct, we don't know the data types of any optional attributes so we just verify the required fields

daniel-kinney-seequent replied:
I'll add a comment here.

daniel-kinney-seequent commented on packages/gef/src/evo/data_converters/gef/converter/gef_to_downhole_collection.py:
```diff
@@ -0,0 +1,439 @@
+        elif len(self.collar_rows) == 1:
+            return self.collar_rows[0]["hole_id"]
+        else:
+            return f"{self.collar_rows[0]['hole_id']}...{self.collar_rows[-1]['hole_id']}"
```
Noting here to review the default name.

daniel-kinney-seequent commented on packages/common/src/evo/data_converters/common/objects/downhole_collection_to_geoscience_object.py:
```diff
@@ -0,0 +1,475 @@
+        arrays = [
+            pa.array(self.dhc.collars.df["final_depth"], type=pa.float64()),
+            pa.array(self.dhc.collars.df["final_depth"], type=pa.float64()),
+            pa.array(self.dhc.collars.df["final_depth"], type=pa.float64()),
```
Is it correct that "final_depth" is used 3 times? If so it's worth a comment, because it looks like it could be a mistake.

daniel-kinney-seequent commented on packages/common/src/evo/data_converters/common/objects/downhole_collection/tables.py:
```diff
@@ -0,0 +1,365 @@
+        if not self._find_column(self.mapping.HOLE_INDEX_COLUMNS):
+            raise ValueError(f"No hole index column found. Expected one of: {self.mapping.HOLE_INDEX_COLUMNS}")
+
+    def _prepare_dataframe(self) -> None:
```
Should this be an abstractmethod?

PaulW8787 replied:
Done.

neilb-sq commented on packages/common/src/evo/data_converters/common/objects/downhole_collection/downhole_collection.py:
```diff
@@ -0,0 +1,145 @@
+        return results
+
+    @override
+    def get_bounding_box(self) -> list[float]:
```
The `BoundingBox_V1_0_1` type, despite not being an intermediary, is probably even less ambiguous. I suspect it's unlikely to change under our feet, and is what Evo will want at the end...

FWIW over in the trimesh space the 2D array of X/Y/Z[Min/Max] is common.

antbrown replied:
I was going by the example in the high level objects by Ben Lewis here, the order of coordinates appears to be expected. Can change though, I did originally have it as a dict

daniel-kinney-seequent replied:
The list is fine, but please review whether `[min_x, min_y, min_z, max_x, max_y, max_z]` is the more approriate ordering.

daniel-kinney-seequent replied:
I am leaving this as is, because it is consistent with the `BoundingBox_V1_0_1` definition and also consistent with other examples of Evo code.

antbrown commented on packages/common/pyproject.toml:
```diff
@@ -16,10 +16,12 @@ dependencies = [
     "requests",
     "scipy",
     "pyproj>=3.7.1",
+    "typing_extensions; python_version < '3.12'",
```
`typing.override` was added in python 3.12 (https://docs.python.org/3/library/typing.html#typing.override) for prior versions we need to install the typing_extensions module to gain access to it

antbrown commented on packages/gef/README.md:
```diff
@@ -0,0 +1,94 @@
+    filepaths=gef_files, evo_workspace_metadata=workspace_metadata, upload_path="gef-cli", overwrite_existing_objects=True
+)
+
+pp(objects_metadata)
```
Confirming this sample code works as expected - so long as you fill in the correct details for client id, hub, redirect etc...

I can publish an object to evo and overwrite it with a new version if I run it again.

If end users don't know their org id or workspace id I can update it to use the same as something like OMF

antbrown commented on packages/common/src/evo/data_converters/common/objects/attributes.py:
```diff
@@ -0,0 +1,266 @@
+        array_class=FloatArray1,
+        attribute_class=ContinuousAttribute,
+        nan_class=NanContinuous,
+    )
```
Yep will remove the factories/classes from this file and make everything module-level :+1:

daniel-kinney-seequent replied:
E: low, P: low

antbrown committed:
Use sync publish objects for now, follow up PR to make async

daniel-kinney-seequent commented on packages/common/src/evo/data_converters/common/objects/attributes.py:
```diff
@@ -0,0 +1,266 @@
+    """
+
+    data_type: DataType
+    array_class: type
```
P: Low, E: Low

daniel-kinney-seequent commented on packages/common/src/evo/data_converters/common/objects/downhole_collection/downhole_collection.py:
```diff
@@ -0,0 +1,145 @@
+    from typing_extensions import override
+
+
+class DownholeCollection(BaseSpatialDataProperties):
```
E: M, P: H. But I'm inclined to leave this to future work.

daniel-kinney-seequent commented on packages/common/src/evo/data_converters/common/objects/downhole_collection/hole_collars.py:
```diff
@@ -0,0 +1,117 @@
+        if not self.is_schema_valid(self.df, HOLE_COLLARS_SCHEMA):
+            raise ValueError("Data is of incorrect type in collars table")
+
+    def is_schema_valid(self, df: pd.DataFrame, schema: dict[str, str]) -> bool:
```
P: low, E: ?

daniel-kinney-seequent replied:
Leaving as is.

daniel-kinney-seequent commented on packages/common/src/evo/data_converters/common/objects/downhole_collection/tables.py:
```diff
@@ -0,0 +1,365 @@
+DEFAULT_DIP: float = 90.0  # Positive dip = down
+
+
+class MeasurementTableAdapter(ABC):
```
E: Low, P: Low

Gary-Lockyer replied:
Done

daniel-kinney-seequent commented on packages/common/src/evo/data_converters/common/objects/downhole_collection/tables.py:
```diff
@@ -0,0 +1,365 @@
+        """
+        self.df: pd.DataFrame = df
+        self.mapping: ColumnMapping = column_mapping
+        self.nan_values_by_column: dict[str, list[typing.Any]] = nan_values_by_column or {}
```
It's a bit of a thinker what to do about this, and the priority is low. One thing to look into is whether the integer columns have a "None" value in the pandas table representation.

daniel-kinney-seequent replied:
I have gone in the opposite direction and removed nan_description from the published objects

daniel-kinney-seequent commented on packages/common/src/evo/data_converters/common/objects/downhole_collection/tables.py:
```diff
@@ -0,0 +1,365 @@
+        by increasing depth within each hole.
+        """
+        self.df.sort_values(
+            by=[self.get_hole_index_column(), self.get_depth_column()], ascending=True, axis=0, inplace=True
```
I'm not sure why it wouldn't already be sorted. I'll catch up with @neilb-sq.

daniel-kinney-seequent replied:
I'm going to trust this is necessary and leave it as is.

daniel-kinney-seequent commented on packages/common/src/evo/data_converters/common/objects/downhole_collection_to_geoscience_object.py:
```diff
@@ -0,0 +1,475 @@
+        coordinate_reference_system = self.create_coordinate_reference_system()
+        bounding_box = self.create_bounding_box()
+
+        distance_unit = UnitLength_UnitCategories("m")
```
Yes, it's changing with Gary's work

daniel-kinney-seequent commented on packages/gef/src/evo/data_converters/gef/converter/gef_to_downhole_collection.py:
```diff
@@ -0,0 +1,439 @@
+            logger.info(f"Using EPSG code {self.epsg_code} from first CPT file")
+        elif self.epsg_code != current_epsg:
+            raise ValueError(
+                f"Inconsistent EPSG codes: {hole_id} has EPSG:{current_epsg}, but expected EPSG:{self.epsg_code}"
```
P: M

daniel-kinney-seequent commented on packages/gef/src/evo/data_converters/gef/converter/gef_to_downhole_collection.py:
```diff
@@ -0,0 +1,439 @@
+        try:
+            _ = cpt_data.delivered_location.x
+            _ = cpt_data.delivered_location.y
+        except AttributeError as e:
```
@reflecting-moon

daniel-kinney-seequent commented on packages/gef/src/evo/data_converters/gef/converter/gef_to_downhole_collection.py:
```diff
@@ -0,0 +1,439 @@
+            k: v
+            for k, v in vars(cpt_data).items()
+            if k not in self.COLLAR_EXCLUDE_KEYS and not k.startswith("_") and not (v is None or v == [] or v == {})
+        }
```
P: H

GaryWL committed:
Add epsg_code parameter to gef_to_evo converter

Add an optional epsg_code parameter which is used to create the
coordinate reference system if the downhole collection does not provide
one.

This also maintains consistency with the other converters

Signed-off-by: Gary Lockyer <gary@catalyst.net.nz>

antbrown committed:
Merge pull request #84 from antbrown/paul-gary-units-gef

Attribute Description records for GEF

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/SeequentEvo/evo-data-converters?pullRequest=124) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/SeequentEvo/evo-data-converters?pullRequest=124) before we can accept your contribution.<br/>**4** out of **5** committers have signed the CLA.<br/><br/>:white_check_mark: reflecting-moon<br/>:white_check_mark: antbrown<br/>:white_check_mark: PaulW8787<br/>:white_check_mark: daniel-kinney-seequent<br/>:x: GaryWL<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/SeequentEvo/evo-data-converters?pullRequest=124) it.</sub>

daniel-kinney-seequent commented on .github/release.yml:
```diff
@@ -8,6 +8,9 @@ changelog:
         - common
+    - title: evo-data-converters-gef
+      labels:
+        - gef
```
Yes you're correct, I was wrong.

GaryWL committed:
Pass the GEF file path

Pass the GEF file path along with the CPTData so that it can be used
in error reporting

Signed-off-by: Gary Lockyer <gary@catalyst.net.nz>

daniel-kinney-seequent commented on packages/gef/src/evo/data_converters/gef/converter/parse_gef_files.py:
```diff
@@ -0,0 +1,135 @@
+    except ValueError:
+        raise
+    except Exception as e:
+        raise RuntimeError(f"Error processing file '{filepath}': {e}") from e
```
P: Low

GaryWL committed:
Rename MeasurementTableAdapter to MeasurementTable

Signed-off-by: Gary Lockyer <gary@catalyst.net.nz>

PaulW8787 committed:
Re-work of DownholeCollection.get_bounding_box() which now computes the boundingbox using the hole depth and trajectory data as well as the collar. Still requires some work for edge-cases and is only tested for DistanceTable data from GEF imports.

GaryWL committed:
Remove MeasurementTableFactory

Signed-off-by: Gary Lockyer <gary@catalyst.net.nz>

antbrown committed:
Merge pull request #86 from Gary-Lockyer/gary-logging

Pass the GEF file path so it can be reported in error messages

Gary-Lockyer commented on packages/common/src/evo/data_converters/common/objects/downhole_collection/tables.py:
```diff
@@ -0,0 +1,365 @@
+    @staticmethod
+    def create(
+        df: pd.DataFrame, column_mapping: ColumnMapping, nan_values_by_column: dict[str, typing.Any] | None = None
+    ) -> MeasurementTableAdapter:
```
Done

antbrown committed:
Merge pull request #88 from Gary-Lockyer/gary-review-clean-ups

Rename MeasurementTableAdapter to MeasurementTable and remove MeasurementTableFactory

antbrown commented on packages/common/src/evo/data_converters/common/objects/downhole_collection/downhole_collection.py:
```diff
@@ -0,0 +1,145 @@
+            self.collars.df["y"].max(),
+            self.collars.df["z"].min(),
+            self.collars.df["z"].max(),
+        ]
```
Paul has done some work on computing the correct bounding box in this PR: https://github.com/antbrown/evo-data-converters/pull/87/changes - I haven't merged it just yet as it may need to take into account the multiple measurement tables that could exist and might need to consider more edge cases.

Let us know if you'd like us to:
* Get someone to pick it up and carry on improving it (Paul works Mon-Wed)
* Merge as-is into gef-feature-branch
* Leave for you or someone else to pick up in the future?

daniel-kinney-seequent replied:
This has been worked on and merged into the feature branch. Interval tables are out of scope.

GaryWL committed:
gef_to_evo add publish_objects parameter

Signed-off-by: Gary Lockyer <gary@catalyst.net.nz>

GaryWL committed:
Remove PyArrowTableFactory make create_table a free function

Signed-off-by: Gary Lockyer <gary@catalyst.net.nz>

GaryWL committed:
Remove AttributeFactory

Signed-off-by: Gary Lockyer <gary@catalyst.net.nz>

GaryWL committed:
Remove complex values from AttributeType Enum

Signed-off-by: Gary Lockyer <gary@catalyst.net.nz>

antbrown committed:
Merge pull request #90 from Gary-Lockyer/gary-cleanups

GEF import review cleanups

daniel-kinney-seequent commented on packages/common/src/evo/data_converters/common/objects/evo_units.txt:
```diff
@@ -0,0 +1,100 @@
+acre_foot_per_barrel = acre_foot/bbl
+barrel_per_acre_foot = bbl/acre_foot
+cubic_foot_per_day_per_barrel_per_day = cubic_foot_per_day/barrel_per_day
+
```
I could use an explanation for why this is in a .txt file.

Gary-Lockyer commented on packages/gef/src/evo/data_converters/gef/importer/gef_to_evo.py:
```diff
@@ -0,0 +1,98 @@
+    from evo.notebooks import ServiceManagerWidget
+
+
+def convert_gef(
```
Done

GaryWL committed:
Pass tags to DownholeCollectionToGeoscienceObject

Manually cherry picked from commits:
    5fa16eacb34875a88f29455c2bd064002ef577fa
    6c416e977743feeec36502c4bb5b40462708f821

    From:   https://github.com/antbrown/evo-data-converters/
    Branch: ags-import-converter-units
    Authored By: Ant Brown

daniel-kinney-seequent commented on packages/common/src/evo/data_converters/common/objects/attributes.py:
```diff
@@ -199,6 +204,23 @@ def create(name: str, series: pd.Series, client: ObjectDataClient) -> OneOfAttri
+            if unit is not None:
+                attribute_description = AttributeDescription(discipline="None", type=unit)
+            else:
+                logger.warning(f"Unable to map {series.dtype} to an EVO unit")
```
Answer: This will result in a non-set attribute description (equivalent to None). This implies we need to extend our mappings that we use to support pint.

GaryWL committed:
Add property white list for dict filtering

Add an attribute white list (COLLAR_ATTRIBUTES) and only extract those
attributes in _get_collar_attributes

Signed-off-by: Gary Lockyer <gary@catalyst.net.nz>

antbrown committed:
Merge pull request #91 from Gary-Lockyer/pass-tags

Pass tags to DownholeCollectionToGeoscienceObject

PaulW8787 committed:
Adds a warning log message if the user tries to convert a GEF fileset in a Python 3.10 environment, advising that GEF additional headers will not be parsed.

GaryWL committed:
Add extra documentation to units

Signed-off-by: Gary Lockyer <gary@catalyst.net.nz>

antbrown committed:
Merge pull request #92 from Gary-Lockyer/property_whitelist

Add property white list for dict filtering

antbrown committed:
Merge pull request #93 from PaulW8787/python-version-3.10-warning

Python 3.10 Warning Message

GaryWL committed:
Rename BaseObjectProperties to BaseGeologicalDataProperties

Signed-off-by: Gary Lockyer <gary@catalyst.net.nz>

GaryWL committed:
Convert attribute comments from """" to #

Signed-off-by: Gary Lockyer <gary@catalyst.net.nz>

GaryWL committed:
Replace for loop with list comprehension

Signed-off-by: Gary Lockyer <gary@catalyst.net.nz>

PaulW8787 committed:
Stricter types on AttributeConfig dataclass.

GaryWL committed:
Clean up exception handling

Signed-off-by: Gary Lockyer <gary@catalyst.net.nz>

GaryWL committed:
Remove redundant hasattr check on data

Signed-off-by: Gary Lockyer <gary@catalyst.net.nz>

antbrown committed:
Merge pull request #94 from PaulW8787/strict-attribute-types

Stricter types on AttributeConfig dataclass

GaryWL committed:
Rename collection_start_end_table collection_interval_table

Signed-off-by: Gary Lockyer <gary@catalyst.net.nz>

PaulW8787 committed:
Change logging level from info to debug on two log messages

antbrown committed:
Merge pull request #95 from PaulW8787/info-to-debug-logmsg-levels

Change logging level from info to debug on two log messages

antbrown committed:
Merge pull request #97 from Gary-Lockyer/document-units

Sundry review clean-ups

PaulW8787 committed:
* Fix code documentation to remove specification that we only cater for vertical holes (re-worded).
* Modify create_dhc_location_attributes() and create_collection_attributes() to return an empty list instead of None, when no attributes are found. NOTE: the former was not mentioned in DKs comment, but is done for consistency.
* Rename arg 'filter' to 'filter_to_table_type' in get_measurement_tables.

antbrown committed:
Merge pull request #98 from PaulW8787/various-minor-fixups-and-tweaks

Various fixes and tweaks

PaulW8787 committed:
Removed some checking for missing Pygef attributes on CPTData

PaulW8787 committed:
* Added a trial type conversion to create_table which guards against erroneous data types sneaking in.
* Also remove duplicate test from test_units, which I noticed in passing.

antbrown committed:
Merge pull request #99 from PaulW8787/remove-some-pygef-attr-checks

Removed some checking for missing Pygef attributes on CPTData

antbrown committed:
Merge pull request #100 from PaulW8787/add-resilience-to-create-table

Add resilience to create table

PaulW8787 committed:
Use @abstractmethod decorator on get_bounding_box method of the BaseSpatialDataProperties abstract class.

PaulW8787 committed:
GEF extract EPSG code method now returns None instead of 'unspecified' when extracted GEF code is 404000.

antbrown committed:
Merge pull request #101 from PaulW8787/abc-method-abstraction

Method abstraction

antbrown committed:
Merge pull request #102 from PaulW8787/epsg-code-unspecified-to-none

EPSG code return value now None

PaulW8787 committed:
Modify the Downhole Collection name to have a prefix of 'GEF CPT {n} holes ' where 'n' is the number of holes. For a single hole the prefix is just 'GEF CPT '.

PaulW8787 committed:
Rename the 'input' argument of DownholeCollection.add_measurement_table() to 'measurements' to avoid shadowing a builtin var name.

antbrown committed:
Merge pull request #103 from PaulW8787/modify-dhc-name

Modify the Downhole Collection name

antbrown committed:
Merge pull request #104 from PaulW8787/avoid-shadowing-input

Avoid shadowing 'input' builtin name

PaulW8787 committed:
Add a warning level log message if more than 1 distance table is found when generating the DownholeCollection geoscience object.

antbrown committed:
Merge pull request #105 from PaulW8787/warn-on-multiple-distance-tables

Warn if more than 1 distance table

antbrown commented on packages/common/src/evo/data_converters/common/objects/downhole_collection/tables.py:
```diff
@@ -0,0 +1,365 @@
+        """
+        Return the name of the primary measurement column.
+
+        :return: Name of the primary column (depth for distance tables, from for interval tables)
```
This may become useful if we want to build multiple location paths and need a single depth value for each measurement in an interval table

daniel-kinney-seequent replied:
I cleaned up `get_primary_columns()`. It wasn't very useful so I removed it from the abstract base class, and just called `get_depth_column()` in the only places that care about it.

PaulW8787 committed:
Add @abstractmethod decorator to MeasurementTable._prepare_dataframe()

PaulW8787 commented on packages/common/src/evo/data_converters/common/objects/base_properties.py:
```diff
@@ -0,0 +1,84 @@
+from abc import ABC
+
+
+class BaseObjectProperties(ABC):
```
Now uses @abstractmethod on get_bounding_box method.

antbrown committed:
Merge pull request #106 from PaulW8787/prepare-dataframe-method-abstract

Abstract method _prepare_dataframe

PaulW8787 commented on packages/common/src/evo/data_converters/common/objects/downhole_collection_to_geoscience_object.py:
```diff
@@ -0,0 +1,475 @@
+
+        :return: Complete location object with spatial and organisational data
+        """
+        dt = self.get_first_distance_measurement_table()
```
Done. Warning log message now added.

daniel-kinney-seequent replied:
AGS import will run into this, so the warning will have to be removed.

daniel-kinney-seequent replied:
I removed it.

daniel-kinney-seequent committed:
Rename helpers with Pandas specific names

daniel-kinney-seequent commented on packages/common/src/evo/data_converters/common/objects/attributes.py:
```diff
@@ -0,0 +1,266 @@
+    )
+
+    # Mapping from inferred pandas dtype to attribute configuration
+    INFERRED_TYPE_MAP: dict[str, AttributeConfig] = {
```
I made this change.

daniel-kinney-seequent commented on packages/common/src/evo/data_converters/common/objects/downhole_collection/column_mapping.py:
```diff
@@ -0,0 +1,51 @@
+
+
+@dataclass
+class ColumnMapping:
```
Leaving as is.

daniel-kinney-seequent committed:
Add comment

daniel-kinney-seequent committed:
Remove warning when there is more than one measurement table

AGS import will run into this, by design.

daniel-kinney-seequent committed:
Cleaned up "get_primary_column"

get_primary_column is not a very useful method. The place where it is used only cares about the depth column of a depth table. Making it generic is just confusing.

get_primary_columns() is a little more useful, so it was kept but renamed.

Removed some tests that do not add value.

daniel-kinney-seequent commented on packages/common/src/evo/data_converters/common/objects/downhole_collection/tables.py:
```diff
@@ -0,0 +1,365 @@
+        return [col for col in self.df.columns if col not in primary]
+
+    @abstractmethod
+    def get_primary_columns(self) -> list[str]:
```
I made this change.

daniel-kinney-seequent committed:
Move some speccy code to gef_spec.py

Also add some comments to help describe more of the GEF spec

daniel-kinney-seequent committed:
Add a test for importing GEFs with different attributes

daniel-kinney-seequent committed:
Add end-to-end tests for GEF importing

These are the first of their kind for the GEF package

daniel-kinney-seequent committed:
Fixup unit test

"predrilled_depth" is not longer being imported as a "collar attribute", but is not being imported from the "raw headers"

daniel-kinney-seequent committed:
Drop nan_description when publishing

It was redundant (and possibly confusing) to publish nan_description sentinels when they have already been converted to NaN

daniel-kinney-seequent committed:
Merge remote-tracking branch 'primary' into gef-feature-branch

daniel-kinney-seequent committed:
Get GEF tests passing

A few of the tests were depending on a `conftest.py` and `consts.py` which didn't exist. They have been re-created.

`test_creates_valid_collection` was broken due to an earlier change, and the coincident attempt to maintain the text had an error.

daniel-kinney-seequent committed:
Merge branch 'gef-feature-branch' into paul-gef-feature-branch-mods

daniel-kinney-seequent committed:
Fix merge conflicts

daniel-kinney-seequent committed:
Add some TODOs for computing the bounding box

daniel-kinney-seequent committed:
Invert vertical/horizontal by swapping cos and sin

daniel-kinney-seequent committed:
Refactor algo to separate function

daniel-kinney-seequent committed:
Don't sort the depths

daniel-kinney-seequent committed:
Handle NaNs and include bbox in test

daniel-kinney-seequent committed:
Temporarily pin python to 3.12

daniel-kinney-seequent committed:
Bump evo dependencies

daniel-kinney-seequent committed:
Temporarily add pytest-asyncio as a dependency

daniel-kinney-seequent committed:
Take a stab at using "typed" evo bindings

I was running into some more issues than I expected with the GEF publishing, requiring me to think about refactoring. I didn't really want to iterate on the current publishing code because I think it has some weaknesses.

There is a "future way" of binding python data classes to Evo in evo-python-sdk. I have taken a stab at publishing a downhole collection this way.

I haven't wired it up to the actual GEF parsing/processing, so right now I only have some tests that demonstrate my bindings could work in principle.

Along the way, I have made a major simplifying data design decision. The "path" part of the DownholeCollection schema supports attributes, and it is enough that the "collections" part of the schema aren't necessary. At least not for the GEF data. This simplifies the data design for this case because there is less data spread out over different tables in the published object.

daniel-kinney-seequent committed:
Merge branch 'paul-gef-feature-branch-mods' into gef-feature-branch

daniel-kinney-seequent committed:
Take a stab at swapping to the typed DHC

Swap from the ad hoc publishing to the "typed" object publishing.

There are a bunch of unfinishined bits.

daniel-kinney-seequent committed:
Make GEF import work again

I mostly reverted my earlier decision to put all of the attributes into the path, based on feedback. I did the "path" related attributes there though, such as depth and trajectory related columns, some of which were used to calculate dip/azimuth. The rest of them went back to the collections, and more typed bindings were written for that.

I copy/paste/edited some of the typed code from evo-python-sdk, because that was the quickest way to experiment and get unblocked. The plan is to push these changes back upstream, probably in an improved manner compared to what I did now.

daniel-kinney-seequent committed:
Make overwrite_existing_objects work

daniel-kinney-seequent committed:
Fix GEF sample

daniel-kinney-seequent committed:
Fixup tests after typed GO refactor

This includes a minor change to rename "depth_veritcal" -> "depth" and "depth" -> "distance" ("penetrationLength" from GEF). This is more consistent with the schema.

daniel-kinney-seequent committed:
Fix publishing for multile holes with different attributes

daniel-kinney-seequent committed:
Remove obsolete code

The refactoring to the "typed" binding for DownholeCollection is complete, so all the old code can be removed.

Also, removed a test which wasn't worth maintaining.

daniel-kinney-seequent committed:
Remove tests for obsolete code removed in previous commit

daniel-kinney-seequent committed:
Fix gef_to_downhole_collection tests

Also, clean up the builder interface to remove unused hole_index paramerter

daniel-kinney-seequent committed:
Merge branch 'refactor-to-typed-object' into gef-feature-branch

daniel-kinney-seequent commented on packages/gef/src/evo/data_converters/gef/objects.py:
```diff
@@ -0,0 +1,559 @@
+    elif isinstance(series.dtype, pd.CategoricalDtype):
+        return "category"
+    elif pd.api.types.is_string_dtype(series):
+        return "string"
```
@BenLewis-Seequent 
It would be more forgiving if we used `pd.api.types.infer_dtype(series, skipna=True)` to sniff out the types.

I had a Series of type "object", something like: `["str1", "str2", pd.NA]`. That didn't pass the `pd.api.types.is_string_dtype` check, so I had to cast it to the type "string". But `pd.api.types.infer_dtype(series, skipna=True)` would have detected that it's a string column.

daniel-kinney-seequent commented on packages/gef/src/evo/data_converters/gef/objects.py:
```diff
@@ -0,0 +1,559 @@
+
+
+assert "date_time" not in _attribute_table_formats
+_attribute_table_formats["date_time"] = [DATE_TIME_ARRAY]
```
This is the most dubious of my upstream "changes".

daniel-kinney-seequent commented:
This PR has old and lots of comments which were addressed. Closed and re-opened: https://github.com/SeequentEvo/evo-data-converters/pull/202
